### ISSUE TITLE
feature request: mode to exit 0 if only autofixes were triggered

### ISSUE THREAD

**Issue by ryanrhee** [pre-commit/pre-commit#2240]
i did a quick search on existing issues but didn't find one that matched what I want. apologies if this is a duplicate.

---

sometimes i find myself running `pre-commit run -a` twice, usually in a script. this most recently happened when i needed to run all my linters as a part of a script. (exact details are described in this issue https://github.com/lerna/lerna/issues/1415#issuecomment-455790478 ; basically, using [lerna](https://github.com/lerna/lerna) to publish a package to npm involves editing a json file and committing the change, and i need to configure a script to run prettier via pre-commit before this commit happens.)

my first try was to run `pre-commit run -a && ... && git add ...`. it did reformat the json file as I wanted, but then the script invocation failed because the call to `pre-commit` exited with a non-zero code. the fix was to change the script invocation to `pre-commit run -a; pre-commit run -a && ... && git add ...`. this makes it to where the first invocation of pre-commit can return any code; assuming that only autofixes were applied, the second invocation should succeed.

i think that when running pre-commit as a generalized lint-runner, the desired behavior changes to the following:

- if everything is green, exit 0
- if things were autofixed, but otherwise everything is ok, still exit 0
- if something is broken and/or not autofix-able, exit nonzero

my feature request is for pre-commit to provide this behavior as a flag, perhaps as a part of the `run` command.

---

anecdotally, i separately remember seeing @mxr 's dotfile, and his alias for `git commit -m "$msg"` was actually something like `git commit -m "$msg" || git commit -m "$msg"` for a similar reason- he basically just wanted to take all pre-commit autofixes as a part of his commit.

---

@asottile i'd like to hear whether this is something you'd be open to having in your project. and if you're open to it, i volunteer @mxr to write the feature

**Comment by asottile** [comment]
I think it's best to let the user perform that scripting if they need it, rather than need to maintain the complexity in the tooling

**Comment by asottile** [comment]
(and 2 might not be enough times -- it may need more times to "settle")

**Comment by ryanrhee** [comment]
Hm yeah good point on needing more than 2 invocations. Thanks for the quick responses!

**Comment by mxr** [comment]
Yeah it's a deliberate design decision that any hook changes require user confirmation, since hooks may not be trustworthy. There are closed issues about this already (I volunteer @ryanrhee to go find them)

[This](https://github.com/mxr/dotfiles/blob/b55fd7d97151d2a77799e7516d24203a71297bcd/.zshrc#L196) is my alias if you decide to trust your hooks 🤷
